prow/plugins/blunderbuss/blunderbuss.go

```diff
@@ -334,34 +360,74 @@ func findReviewer(ghc githubClient, log *logrus.Entry, useStatusAvailability boo
 			// we've already verified this reviewer is busy
 			continue
 		}
-		busy, err := isUserBusy(ghc, candidate)
-		if err != nil {
-			log.Errorf("error checking user availability: %v", err)
-		}
-		if !busy {
-			return candidate
-		}
-		// if we haven't returned the candidate, then they must be busy.
-		busyReviewers.Insert(candidate)
+		return candidate
 	}
 	return ""
 }
 
-type githubAvailabilityQuery struct {
-	User struct {
-		Login  githubql.String
-		Status struct {
-			IndicatesLimitedAvailability githubql.Boolean
+// prepareLeafReviewersToQuery prepares reviewers list using info from reviewers client crossed with author set
+func prepareLeafReviewersToQuery(rc reviewersClient, ghc githubClient, files []github.PullRequestChange, authorSet sets.String) []string {
+	ownersSeen := sets.NewString()
+	reviewersToQuery := layeredsets.NewString()
+	for _, file := range files {
+		ownersFile := rc.FindReviewersOwnersForFile(file.Filename)
+		if ownersSeen.Has(ownersFile) {
+			continue
 		}
-	} `graphql:"user(login: $user)"`
+		ownersSeen.Insert(ownersFile)
+		reviewers := layeredsets.NewString(rc.LeafReviewers(file.Filename).List()...).Difference(authorSet)
+		reviewersToQuery = reviewersToQuery.Union(reviewers)
+	}
+	return reviewersToQuery.List()
 }
 
-func isUserBusy(ghc githubClient, user string) (bool, error) {
-	var query githubAvailabilityQuery
-	vars := map[string]interface{}{
-		"user": githubql.String(user),
+type githubUserAvailability struct {
+	Login  string
+	Status struct {
+		IndicatesLimitedAvailability bool
+	}
+}
+
+// graphqlTag helps in formating the graphql tag
+func graphqlTag(value string) reflect.StructTag {
+	return reflect.StructTag(strconv.AppendQuote([]byte("graphql:"), value))
+}
+
+// findBusyReviewers finds all the people that are busy and excluded from the review
+func findBusyReviewers(ghc githubClient, log *logrus.Entry, busyReviewers *sets.String, reviewers []string) {
+	const (
+		// will panic if too many fields with too much data in the reflect struct, this setting is very safe
+		githubUserAvailabilityStructLimit = 300
```

Comment:
When huge reviewers set is queried, users are divided into batches. Batches have to have limits due to too big structure error in golang or too much data problem. This limit is safe. Allows asking for users where every user has more than 100 chars in its login.
This is introduced to prevent panic in heavy PRs:
- PR: https://github.com/openshift/cluster-authentication-operator/pull/517 asked for more then 1600 reviewers. In the current situation, it got answered around 30 times and the rest was unchecked due to rate limit error
- Another similar example is PR: https://github.com/openshift/operator-framework-olm/pull/219 more than 1400 single queries about the user.
Without division to batches, reflect would panic due to too large structure.